In [1]:
results = {}

# Int

- Memory: Min size 24 bits. Max size is not limited.
- Performance: fixed object creation speed (11 ns)

In [2]:
import sys

dtype = int

value = 0
print('{}({}): {} bits'.format(str(dtype.__name__), value, sys.getsizeof(value)))
for i in [0, 1, 128, 512, 1023, 2048]:
    value = dtype(2 ** i)
    print('{}({} ** {}): {} bits'.format(str(dtype.__name__), 2, i, sys.getsizeof(value)))

int(0): 24 bits
int(2 ** 0): 28 bits
int(2 ** 1): 28 bits
int(2 ** 128): 44 bits
int(2 ** 512): 96 bits
int(2 ** 1023): 164 bits
int(2 ** 2048): 300 bits


In [3]:
v0 = %timeit -o t = 0
v1 = %timeit -o t = 1000
v2 = %timeit -o t = 1000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000

results['int'] = [v0.average, v1.average, v2.average]

11.9 ns ± 0.0454 ns per loop (mean ± std. dev. of 7 runs, 100000000 loops each)
11.1 ns ± 0.0871 ns per loop (mean ± std. dev. of 7 runs, 100000000 loops each)
10.8 ns ± 0.0525 ns per loop (mean ± std. dev. of 7 runs, 100000000 loops each)


In [4]:
%timeit t = int(0)

92.2 ns ± 1.14 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


# Float

- Memory: Min and max size 24 bits.
- Performance: fixed object creation speed (11 ns). The same as int.

In [5]:
import sys

dtype = float

value = 0
print('{}({}): {} bits'.format(str(dtype.__name__), value, sys.getsizeof(value)))
for i in [0, 1, 128, 512, 1023, 2048]:
    try:
        value = dtype(2 ** i)
    except OverflowError as e:
        print("Exception {}({} ** {}). {}".format(str(dtype.__name__), 2, i, e))
        continue
    print('{}({} ** {}): {} bits'.format(str(dtype.__name__), 2, i, sys.getsizeof(value)))

float(0): 24 bits
float(2 ** 0): 24 bits
float(2 ** 1): 24 bits
float(2 ** 128): 24 bits
float(2 ** 512): 24 bits
float(2 ** 1023): 24 bits
Exception float(2 ** 2048). int too large to convert to float


In [6]:
# max exp = 1024
sys.float_info

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)

In [7]:
v0 = %timeit -o t = 0.
v1 = %timeit -o t = 1000.
v2 = %timeit -o t = 1000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000.

results['float'] = [v0.average, v1.average, v2.average]

10.9 ns ± 0.0909 ns per loop (mean ± std. dev. of 7 runs, 100000000 loops each)
10.9 ns ± 0.0548 ns per loop (mean ± std. dev. of 7 runs, 100000000 loops each)
11 ns ± 0.0667 ns per loop (mean ± std. dev. of 7 runs, 100000000 loops each)


In [8]:
%timeit t = float(0.)

95.1 ns ± 2.12 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


# Complex

- Memory: Min and max size 32 bits.
- Performance: fixed object creation speed (157 ns). 15x slower than int.

In [9]:
import sys

dtype = complex

value = complex(0, 0)
print('{}({}): {} bits'.format(str(dtype.__name__), value, sys.getsizeof(value)))
for i in [0, 1, 128, 512, 1023, 2048]:
    try:
        value = dtype(2 ** i, 2 ** -i)
    except OverflowError as e:
        print("Exception {}({} ** {}). {}".format(str(dtype.__name__), 2, i, e))
        continue
    print('{}({} ** {}): {} bits'.format(str(dtype.__name__), 2, i, sys.getsizeof(value)))

complex(0j): 32 bits
complex(2 ** 0): 32 bits
complex(2 ** 1): 32 bits
complex(2 ** 128): 32 bits
complex(2 ** 512): 32 bits
complex(2 ** 1023): 32 bits
Exception complex(2 ** 2048). int too large to convert to float


In [10]:
v0 = %timeit -o t = complex(0., 0.)
v1 = %timeit -o t = complex(1000., 0.0001)
v2 = %timeit -o t = complex(1000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000., 0.000000001)

results['complex'] = [v0.average, v1.average, v2.average]

161 ns ± 0.899 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
162 ns ± 0.946 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
165 ns ± 4.72 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


# Str

- Memory: Min size 49 bits. Max size is not limited (length of string plus 49 bits).
- Performance: fixed object creation speed (11 ns). The same as int.

In [11]:
import sys

dtype = str

value = ''
print('{}({}): {} bits'.format(str(dtype.__name__), value, sys.getsizeof(value)))
for i in [0, 1, 128, 512, 1023, 2048]:
    value = dtype(2 ** i)
    print('{}({} ** {}): {} bits (string length {} + 49)'.format(str(dtype.__name__), 2, i, sys.getsizeof(value), len(value)))

str(): 49 bits
str(2 ** 0): 50 bits (string length 1 + 49)
str(2 ** 1): 50 bits (string length 1 + 49)
str(2 ** 128): 88 bits (string length 39 + 49)
str(2 ** 512): 204 bits (string length 155 + 49)
str(2 ** 1023): 357 bits (string length 308 + 49)
str(2 ** 2048): 666 bits (string length 617 + 49)


In [12]:
v0 = %timeit -o t = ''
v1 = %timeit -o t = '1000'
v2 = %timeit -o t = '1000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000'

results['str'] = [v0.average, v1.average, v2.average]

10.9 ns ± 0.104 ns per loop (mean ± std. dev. of 7 runs, 100000000 loops each)
11.1 ns ± 0.0248 ns per loop (mean ± std. dev. of 7 runs, 100000000 loops each)
11 ns ± 0.0709 ns per loop (mean ± std. dev. of 7 runs, 100000000 loops each)


In [13]:
%timeit t = str('')

95 ns ± 0.317 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


# List

- Memory: Min size 64 bits. Max size is not limited.
- Performance: Min wall time 34 ns (3x slower than int). Object creation speed grows linearly.

In [14]:
import sys

dtype = list

value = []
print('{}({}): {} bits'.format(str(dtype.__name__), value, sys.getsizeof(value)))
for i in [0, 1, 128, 512, 1023, 2048]:
    value = dtype([x for x in range(i)])
    print('{}(elements = {}): {} bits'.format(str(dtype.__name__), i, sys.getsizeof(value), len(value)))

list([]): 64 bits
list(elements = 0): 64 bits
list(elements = 1): 96 bits
list(elements = 128): 1264 bits
list(elements = 512): 4720 bits
list(elements = 1023): 9312 bits
list(elements = 2048): 18544 bits


In [15]:
v0 = %timeit -o t = [0, 1]
v1 = %timeit -o t = [0, 1, 2, 3, 4, 5, 6, 7]
v2 = %timeit -o t = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15,\
                     16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31]

results['list'] = [v0.average, v1.average, v2.average]

36.5 ns ± 0.403 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
62.2 ns ± 0.554 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
176 ns ± 2.14 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


In [16]:
%timeit t = list([])

125 ns ± 1.01 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


# Tuple

- Memory: Min size 48 bits. Max size is not limited.
- Performance: Fixed object creation speed (11 ns). The same as int. (Awesome!)

In [17]:
import sys

dtype = tuple

value = ()
print('{}({}): {} bits'.format(str(dtype.__name__), value, sys.getsizeof(value)))
for i in [0, 1, 128, 512, 1023, 2048]:
    value = dtype((x for x in range(i)))
    print('{}(elements = {}): {} bits'.format(str(dtype.__name__), i, sys.getsizeof(value), len(value)))

tuple(()): 48 bits
tuple(elements = 0): 48 bits
tuple(elements = 1): 56 bits
tuple(elements = 128): 1072 bits
tuple(elements = 512): 4144 bits
tuple(elements = 1023): 8232 bits
tuple(elements = 2048): 16432 bits


In [18]:
v0 = %timeit -o t = (0, 1)
v1 = %timeit -o t = (0, 1, 2, 3, 4, 5, 6, 7)
v2 = %timeit -o t = (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15,\
                     16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31)

results['tuple'] = [v0.average, v1.average, v2.average]

11 ns ± 0.109 ns per loop (mean ± std. dev. of 7 runs, 100000000 loops each)
12.2 ns ± 0.975 ns per loop (mean ± std. dev. of 7 runs, 100000000 loops each)
11.3 ns ± 0.279 ns per loop (mean ± std. dev. of 7 runs, 100000000 loops each)


In [19]:
%timeit t = tuple()

74.9 ns ± 2.8 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


# Dict

- Memory: Min size 240 bits. Max size is not limited.
- Performance: Min wall time 64ns (2x slower than list). Object creation speed grows linearly. (4x slower than list)

In [20]:
import sys

dtype = dict

value = {}
print('{}({}): {} bits'.format(str(dtype.__name__), value, sys.getsizeof(value)))
for i in [0, 1, 128, 512, 1023, 2048]:
    value = dtype({x: x for x in range(i)})
    print('{}(elements = {}): {} bits'.format(str(dtype.__name__), i, sys.getsizeof(value), len(value)))

dict({}): 240 bits
dict(elements = 0): 240 bits
dict(elements = 1): 240 bits
dict(elements = 128): 4704 bits
dict(elements = 512): 18528 bits
dict(elements = 1023): 36968 bits
dict(elements = 2048): 73824 bits


In [21]:
v0 = %timeit -o t = {0: 0, 1: 1}
v1 = %timeit -o t = {0: 0, 1: 1, 2: 2, 3: 3, 4: 4, 5: 5, 6: 6, 7: 7}
v2 = %timeit -o t = {0: 0, 1: 1, 2: 2, 3: 3, 4: 4, 5: 5, 6: 6, 7: 7, 8: 8, 9: 9, 10: 10,\
                     11: 11, 12: 12, 13: 13, 14: 14, 15: 15, 16: 16, 17: 17, 18: 18, 19: 19, 20: 20,\
                     21: 21, 22: 22, 23: 23, 24: 24, 25: 25, 26: 26, 27: 27, 28: 28, 29: 29, 30: 30,\
                     31: 31}

results['dict'] = [v0.average, v1.average, v2.average]

84.2 ns ± 0.524 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
225 ns ± 2.26 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
777 ns ± 8.73 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [22]:
%timeit t = dict()

93.4 ns ± 0.694 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


# Set

- Memory: Min memory size 224 bits. Max size is not limited.
- Performance: Min wall time 52 ns (1.5x slower than list). Object creation speed grows linearly. (2.5x slower than list)

In [23]:
import sys

dtype = set

value = set()
print('{}({}): {} bits'.format(str(dtype.__name__), value, sys.getsizeof(value)))
for i in [0, 1, 128, 512, 1023, 2048]:
    value = dtype({x for x in range(i)})
    print('{}(elements = {}): {} bits'.format(str(dtype.__name__), i, sys.getsizeof(value), len(value)))

set(set()): 224 bits
set(elements = 0): 224 bits
set(elements = 1): 224 bits
set(elements = 128): 8416 bits
set(elements = 512): 32992 bits
set(elements = 1023): 32992 bits
set(elements = 2048): 131296 bits


In [24]:
v0 = %timeit -o t = {0, 1}
v1 = %timeit -o t = {0, 1, 2, 3, 4, 5, 6, 7}
v2 = %timeit -o t = {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15,\
                     16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31}

results['set'] = [v0.average, v1.average, v2.average]

60.3 ns ± 0.327 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
166 ns ± 0.339 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
626 ns ± 7.46 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


In [25]:
%timeit t = set({0})

141 ns ± 0.712 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


# Frozenset

- Memory: Min memory size 224 bits. Max size is not limited.
- Performance: Min wall time 159 ns (5x slower than list). Object creation speed grows linearly. (4.5x slower than list)

In [26]:
import sys

dtype = frozenset

value = frozenset()
print('{}({}): {} bits'.format(str(dtype.__name__), value, sys.getsizeof(value)))
for i in [0, 1, 128, 512, 1023, 2048]:
    value = dtype({x for x in range(i)})
    print('{}(elements = {}): {} bits'.format(str(dtype.__name__), i, sys.getsizeof(value), len(value)))

frozenset(frozenset()): 224 bits
frozenset(elements = 0): 224 bits
frozenset(elements = 1): 224 bits
frozenset(elements = 128): 8416 bits
frozenset(elements = 512): 32992 bits
frozenset(elements = 1023): 32992 bits
frozenset(elements = 2048): 131296 bits


In [27]:
v0 = %timeit -o t = frozenset((0, 1))
v1 = %timeit -o t = frozenset((0, 1, 2, 3, 4, 5, 6, 7))
v2 = %timeit -o t = frozenset((0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15,\
                               16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31))

results['frozenset'] = [v0.average, v1.average, v2.average]

146 ns ± 0.878 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
235 ns ± 2.38 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
628 ns ± 5.34 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


# Decimal

- Memory: Min 104 bits. Max size is not limited.
- Performance: Min wall time 134 ns (13x slower than int). Object creation speed grows exponentionaly from the value of object. (150x slower than int)


In [28]:
import sys

from decimal import Decimal

dtype = Decimal

value = Decimal(0)
print('{}({}): {} bits'.format(str(dtype.__name__), value, sys.getsizeof(value)))
for i in [0, 1, 128, 512, 1023, 2048]:
    value = dtype(2 ** i)
    print('{}({} ** {}): {} bits'.format(str(dtype.__name__), 2, i, sys.getsizeof(value)))

Decimal(0): 104 bits
Decimal(2 ** 0): 104 bits
Decimal(2 ** 1): 104 bits
Decimal(2 ** 128): 104 bits
Decimal(2 ** 512): 176 bits
Decimal(2 ** 1023): 240 bits
Decimal(2 ** 2048): 368 bits


In [29]:
v0 = %timeit -o t = Decimal(10.)
v1 = %timeit -o t = Decimal(10000000.)
v2 = %timeit -o t = Decimal(1000000000000000000000000000000000000000000000000000000000000000000000000000.)

results['Decimal'] = [v0.average, v1.average, v2.average]

405 ns ± 3.96 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
448 ns ± 5.81 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
1.18 µs ± 2.92 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


# Defaultdict

- Memory: Min size 248 bits. Max size is not limited.
- Performance: Min wall time 205ns (8x slower than list). Object creation speed grows linearly from the number of objects. (9x slower than list)

In [30]:
import sys
from collections import defaultdict

dtype = defaultdict

value = dtype(int)
print('{}(): {} bits'.format(str(dtype.__name__), sys.getsizeof(value)))
for i in [0, 1, 128, 512, 1023, 2048]:
    value = dtype(int)
    for x in range(i):
        value[x] += 1
    print('{}(elements = {}): {} bits'.format(str(dtype.__name__), i, sys.getsizeof(value)))

defaultdict(): 248 bits
defaultdict(elements = 0): 248 bits
defaultdict(elements = 1): 248 bits
defaultdict(elements = 128): 4712 bits
defaultdict(elements = 512): 18536 bits
defaultdict(elements = 1023): 36976 bits
defaultdict(elements = 2048): 73832 bits


In [31]:
%%timeit -o

t = defaultdict(int)
for k in (0, 1):
    t[k] += k

585 ns ± 4.35 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


<TimeitResult : 585 ns ± 4.35 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)>

In [32]:
v0 = _

In [33]:
%%timeit -o

t = defaultdict(int)
for k in (0, 1, 2, 3, 4, 5, 6, 7):
    t[k] += k

1.69 µs ± 23.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


<TimeitResult : 1.69 µs ± 23.9 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)>

In [34]:
v1 = _

In [35]:
%%timeit -o

t = defaultdict(int)
for k in (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15,
          16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31):
    t[k] += k

6.04 µs ± 100 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)


<TimeitResult : 6.04 µs ± 100 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)>

In [36]:
v2 = _

In [37]:
results['defaultdict'] = [v0.average, v1.average, v2.average]

# Class instance

- Memory: Min size 168 bits.
- Performance: Wall time 235 ns. 20x slower than int.

In [38]:
class Point:
    
    def __init__(self, x):
        self.x = x
        
ob = Point(1)

print('{}: {} bits'.format(str(ob.__class__.__name__), sys.getsizeof(ob) + sys.getsizeof(ob.__dict__)))

Point: 168 bits


In [39]:
v0 = %timeit -o t = Point(1)

results['Class'] = [v0.average, None, None]

227 ns ± 10.1 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


# Namedtuple

- Memory: Min size 56 bits. (3x less memory than class)
- Performance: Wall time 308 ns. 27x slower than int.

In [40]:
from collections import namedtuple

Point = namedtuple('Point', ('x'))

ob = Point(1)

print('{}: {} bits'.format(str(ob.__class__.__name__), sys.getsizeof(ob)))

Point: 56 bits


In [41]:
v0 = %timeit -o t = Point(1)

results['namedtuple'] = [v0.average, None, None]

313 ns ± 7.6 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)


# Class instance with slots

- Memory: Min size 48 bits (3.5x less memory than class).
- Performance: Wall time 194 ns. (20% faster than class). 17x sower than int.

In [42]:
class Point:
    __slots__ = ['x']
    
    def __init__(self, x):
        self.x = x
        
ob = Point(1)

print('{}: {} bits'.format(str(ob.__class__.__name__), sys.getsizeof(ob)))

Point: 48 bits


In [43]:
v0 = %timeit -o t = Point(1)

results['ClassWithSlots'] = [v0.average, None, None]

185 ns ± 5.27 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


# Performance table

In [44]:
import pandas as pd

df = (pd.DataFrame(results, index=['test 0', 'test 1', 'test 2']) * 1e9).round(1).T
df

test 0  test 1  test 2
int               11.9    11.1    10.8
float             10.9    10.9    11.0
complex          161.2   161.7   165.3
str               10.9    11.1    11.0
list              36.5    62.2   175.8
tuple             11.0    12.2    11.3
dict              84.2   225.5   776.6
set               60.3   166.4   626.1
frozenset        145.9   234.6   628.1
Decimal          405.4   447.7  1182.8
defaultdict      585.1  1690.1  6044.5
Class            227.3     NaN     NaN
namedtuple       313.1     NaN     NaN
ClassWithSlots   184.6     NaN     NaN

# Relative performance table 

(should by 4x when linear)

In [45]:
df_ratio = df.copy(True)
df_ratio['test 2'] = df_ratio['test 2'] / df_ratio['test 1']
df_ratio['test 1'] = df_ratio['test 1'] / df_ratio['test 0']
df_ratio['test 0'] = 1.0

df_ratio = df_ratio.round(2)
df_ratio

test 0  test 1  test 2
int                1.0    0.93    0.97
float              1.0    1.00    1.01
complex            1.0    1.00    1.02
str                1.0    1.02    0.99
list               1.0    1.70    2.83
tuple              1.0    1.11    0.93
dict               1.0    2.68    3.44
set                1.0    2.76    3.76
frozenset          1.0    1.61    2.68
Decimal            1.0    1.10    2.64
defaultdict        1.0    2.89    3.58
Class              1.0     NaN     NaN
namedtuple         1.0     NaN     NaN
ClassWithSlots     1.0     NaN     NaN

# Types with constant initialization time

In [46]:
df.loc[['int', 'float', 'complex', 'str', 'tuple']]

test 0  test 1  test 2
int        11.9    11.1    10.8
float      10.9    10.9    11.0
complex   161.2   161.7   165.3
str        10.9    11.1    11.0
tuple      11.0    12.2    11.3

# Types with linear initialization time

In [51]:
df.loc[['list', 'dict', 'set', 'frozenset', 'defaultdict']]

test 0  test 1  test 2
list           36.5    62.2   175.8
dict           84.2   225.5   776.6
set            60.3   166.4   626.1
frozenset     145.9   234.6   628.1
defaultdict   585.1  1690.1  6044.5

In [52]:
df_ratio.loc[['list', 'dict', 'set', 'frozenset', 'defaultdict']]

test 0  test 1  test 2
list            1.0    1.70    2.83
dict            1.0    2.68    3.44
set             1.0    2.76    3.76
frozenset       1.0    1.61    2.68
defaultdict     1.0    2.89    3.58

# Types with unpredictable initialization time

In [48]:
df.loc[['Decimal']]

test 0  test 1  test 2
Decimal   405.4   447.7  1182.8